### ISSUE TITLE
The C-like-shader-source backends have lots of common code that should be shared

### ISSUE THREAD

**Issue by abadams** [halide/Halide#4977]
E.g. mapping types into a subset supported by GLSL/HLSL/Vulkan, finding shared memory allocations.

**Comment by zvookin** [comment]
Yes. One design pattern that is fairly productive is to structure these as passes. E.g. lowering a more general set of types into a restricted one. Allocation transforms would also likely be good as passes. (I also want more scheduling control over this, but that is a harder design problem.)

I'm less enthusiastic about trying to make common code via an API for some of this. There are cases where it is a good idea, but it tends to be fraught with tradeoffs such that either the API gets too complicated or the functionality assumes too much about a specific backend. That said, there are cases where a common code function would help.
